kubetest/gcedind/gce_dind.go

```diff
@@ -0,0 +1,622 @@
+	envMap := map[string]string{
+		"NUM_NODES":                   strconv.Itoa(d.numNodes),
+		"KUBE_DIND_GCE_PROJECT":       d.project,
+		"KUBE_DIND_GCE_ZONE":          d.zone,
+		"DIND_IMAGE":                  d.dindImage,
+		"BUILD_KUBEADM":               buildKubeadm,
+		"BUILD_HYPERKUBE":             buildHyperkube,
+		"IP_MODE":                     ipMode,
+		"KUBERNETES_CONFORMANCE_TEST": "yes",
+	}
+	for env, val := range envMap {
+		if err := os.Setenv(env, val); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// dockerMachineConnect sets a docker-machine connection from the local
+// host to the GCE instance. It is equivalent to setting environment
+// variables via 'eval $(docker-machine env k8s-dind)'.
+func (d *Deployer) dockerMachineConnect() error {
+	o, err := d.getOutput("docker-machine env k8s-dind")
+	if err != nil {
+		return err
+	}
+	trimmed := strings.TrimSpace(string(o))
+	if trimmed != "" {
+		lines := strings.Split(trimmed, "\n")
+		for _, line := range lines {
+			// If the output line is of the form:
+			//     export KEY="VALUE"
+			// then trim the leading "export ", remove the quotes, and split
+			// into KEY and VALUE fields
+			if strings.Contains(line, "export") {
+				keyPair := strings.Split(strings.Replace(strings.TrimPrefix(line, "export "), "\"", "", -1), "=")
+				os.Setenv(keyPair[0], keyPair[1])
+			}
+		}
+	}
+	return nil
+}
+
+// clusterSize determines the number of nodes in a cluster.
+func (d *Deployer) clusterSize() (int, error) {
+	o, err := d.getOutput("kubectl get nodes --no-headers")
+	if err != nil {
+		return -1, fmt.Errorf("kubectl get nodes failed: %s\n%s", err, string(o))
+	}
+	trimmed := strings.TrimSpace(string(o))
+	if trimmed != "" {
+		return len(strings.Split(trimmed, "\n")), nil
+	}
+	return 0, nil
+}
+
+// checkProject confirms that the expected GCE project exists.
+func (d *Deployer) checkProject() error {
+	log.Printf("Confirming GCE project '%s'", d.project)
+	o, err := d.getOutput("gcloud config list project --format value(core.project)")
+	if err != nil {
+		return err
+	}
+	if project := strings.TrimSpace(string(o)); project != d.project {
+		return fmt.Errorf("GCE project is '%s', expected project '%s'", project, d.project)
+	}
+	return nil
+}
+
+// checkInstance confirms that the expected GCE instance exists.
+func (d *Deployer) checkInstance() error {
+	log.Printf("Confirming GCE instance '%s'", gceInstanceName)
+	cmd := fmt.Sprintf("gcloud compute instances list --filter=name=(%s)", gceInstanceName)
+	o, err := d.getOutput(cmd)
+	if err != nil {
+		return err
+	}
+	if !strings.Contains(string(o), gceInstanceName) {
+		return fmt.Errorf("Could not find instance '%s'", gceInstanceName)
+	}
+	return nil
+}
+
+// saveInstanceLogs collects a serial log, service logs, and docker state
+// from the GCE instance, and saves the logs in a local artifacts directory.
+func (d *Deployer) saveInstanceLogs(artifactsDir string) error {
```

Comment:
This functionality is more general than the dind tests. Could you move this code into another package, so we can later use it to break the general gce e2e test dependency on the log recovery bash script in k/k?

Reply:
I'll take a look at how this can be factored out. Just to make sure we're on the same page...
Is the functionality that you're discussing moving out just for the GCE instance, or for both GCE instance + Master/Worker nodes (i.e. functionality that would apply to any setup that has master/worker nodes that are reachable via 'docker exec <container> <cmd>')?

If I move the GCE instance logging functionality to a new (general purpose) package, it would have to be a little bit flexible in allowing the caller/user to provide lists of what to collect. For example, in my specific GCE DinD case, the one GCE instance won't be doing kube-api, kube-controller, kube-proxy, etc., but in other GCE cases, a GCE master instance will be running these system pods. (Probably not a big deal, just argument passing).

For the master/worker log collection, I'm using 'docker exec <container> <cmd>' to collect logs (this works because of the docker-machine connection to the instance/VM), but this might also work for other cases, e.g. the local DinD case, I believe.

Reply:
We should probably take those lists as formal parameters anyway, or at least from member fields.

I just want the log scraping for a gce node moved out. The bash script we currently use in the e2e tests is functionally similar to this, but has deep import dependencies on the k/k/cluster dir. For pods, I suspect we'll be able to just scrape the logs from the mirror pods via the k8s api.

I'll read through this again tomorrow and check that I'm making sense.

Reply:
I took a look at how the GCE instance log collecting can be factored out separately in a general-purpose way that will work for other GCE-based CIs. I think this is a great idea, but I'm wondering if this should be done as a separate, followup change. My thought process is as follows: I started by looking at kubetest/dump.go (used in the kops-based CI tests) as a model for what might be needed. (The dump.go relies on the user providing an SSH client factory for creating client connections for each node VM. Perhaps GCE-base CIs could re-use this dump.go framework with a standard SSH client, since 'gcloud compute ssh ...' is supposedly a wrapper around ssh???) Using dump.go as a model, this framework uses 'kubectl get nodes' and then calls the d.dumpNode() method to collect the logs for each node in turn. So let's say that I do go ahead and add a new "gcedump" package to provide similar general-purpose functionality for GCE-based CI. Most of the logic in this new package wouldn't get exercised in the gce-DinD case. For example, I would have to tell the "gcedump" to ignore the 'kubectl get nodes' output (since kubernetes nodes are containers, not VMs, in gce-DinD), and I'd have to tell it to also collect logs for the GCE instance in this special case. So in order to test this new "gcedump" framework, I would also have to test it on an existing, more typical GCE-based CI test job. This would be development work that's really outside the scope of what I'm trying to add for GCE-DinD. Or I could just add a package that only collects logs for a single GCE VM, but ultimately I think we want something that does the log collection for all nodes on GCE-based tests. Your thoughts?